Banjo-Tooie: implement new game

jjjj12212 opened the pull request:
## What is this fixing or adding?
This is adding a new Archipelago game that was in development since November 2023. It has come a long way with lots of features and still more features to come in the future. We believe that in its current state, it is now suitable to be released officially.

## How was this tested?
The game was posted in the future-game-design since February 2024 and since then has received feedback & testing from the community. Bug reports are submitted from the community and issues has been resolved between versions.

## If this makes graphical changes, please attach screenshots.
If nestanity is enabled, nests throughout the game will be displayed with the archipelago logo.
![archipelago_nests](https://github.com/user-attachments/assets/bf04284b-3b8d-4dd5-8b67-6da9043f7181)

jjjj12212 committed:
Banjo-Tooie AP V4.1

jjjj12212 committed:
Update Documentation for AP Release

g0goTAS committed:
French docs, oui oui poutine omelette du fromage

g0goTAS committed:
wording fix

jjjj12212 committed:
logic and placement fixes + trap weights

jjjj12212 committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        if (item.code == self.item_code(itemName.TTRAP) or item.code == self.item_code(itemName.STRAP) or item.code == self.item_code(itemName.TRTRAP) or item.code == self.item_code(itemName.SQTRAP)) and self.options.traps.value == False:
+            return False
+        
+        if self.options.progressive_beak_buster.value == True and (item.code == self.item_code(itemName.BBUST) or item.code == self.item_code(itemName.BDRILL)):
```
```suggestion
        if self.options.progressive_beak_buster and (item.code == self.item_code(itemName.BBUST) or item.code == self.item_code(itemName.BDRILL)):
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        if self.options.progressive_beak_buster.value == True and (item.code == self.item_code(itemName.BBUST) or item.code == self.item_code(itemName.BDRILL)):
+            return False
+        
+        if item.code == self.item_code(itemName.PBBUST) and self.options.progressive_beak_buster.value == False:
+            return False
+        
+        if item.code == self.item_code(itemName.PFLIGHT) and self.options.progressive_flight.value == False:
```
```suggestion
        if item.code == self.item_code(itemName.PBBUST) and not self.options.progressive_beak_buster:
            return False
        
        if item.code == self.item_code(itemName.PFLIGHT) and not self.options.progressive_flight:
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        if self.options.egg_behaviour.value == 1 and item.code == self.starting_egg: #Already has this egg in inventory
+            return False
+        
+        if self.options.progressive_shoes.value == True and (item.code == self.item_code(itemName.SSTRIDE) or item.code == self.item_code(itemName.TTRAIN) \
```
```suggestion
        if self.options.progressive_shoes and (item.code == self.item_code(itemName.SSTRIDE) or item.code == self.item_code(itemName.TTRAIN) \
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        if self.options.progressive_shoes.value == True and (item.code == self.item_code(itemName.SSTRIDE) or item.code == self.item_code(itemName.TTRAIN) \
+            or item.code == self.item_code(itemName.SPRINGB) or item.code == self.item_code(itemName.CLAWBTS)):
+            return False
+        if item.code == self.item_code(itemName.PSHOES) and self.options.progressive_shoes.value == False:
```
```suggestion
        if item.code == self.item_code(itemName.PSHOES) and not self.options.progressive_shoes:
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        if self.options.progressive_egg_aiming.value == 2 and (item.code == self.item_code(itemName.AMAZEOGAZE) or item.code == self.item_code(itemName.BBLASTER)):
+            return False
+        
+        if self.options.progressive_bash_attack.value == True and (item.code == self.item_code(itemName.BBASH) or item.code == self.item_code(itemName.GRAT)):
+            return False
+        if item.code == self.item_code(itemName.PBASH) and self.options.progressive_bash_attack.value == False:
+            return False
+        
+        if item.code == self.item_code(itemName.ROAR) and self.options.randomize_dino_roar.value == False:
```
```suggestion
        if self.options.progressive_bash_attack and (item.code == self.item_code(itemName.BBASH) or item.code == self.item_code(itemName.GRAT)):
            return False
        if item.code == self.item_code(itemName.PBASH) and not self.options.progressive_bash_attack:
            return False
        
        if item.code == self.item_code(itemName.ROAR) and not self.options.randomize_dino_roar:
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        self.pre_fill_me()
+
+    def generate_early(self) -> None:
+        if self.options.randomize_worlds.value == True and self.options.randomize_bk_moves.value != 0 and self.options.logic_type == 0:
+            raise ValueError("Randomize Worlds and Randomize BK Moves is not compatible with Beginner Logic.")
+        if self.options.randomize_notes == False and self.options.randomize_worlds.value == True and self.options.randomize_bk_moves.value != 0:
+            if self.multiworld.players == 1:
+                raise ValueError("Randomize Notes is required for Randomize BK Moves and Randomize Worlds enabled.")
+        if self.options.randomize_notes == False and (self.options.extra_trebleclefs_count.value != 0 and self.options.bassclef_amount.value != 0):
+            raise ValueError("Randomize Notes is required to add extra Treble Clefs or Bass Clefs")
+        if self.options.progressive_beak_buster.value == True and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive Beak Buster without randomizing moves and randomizing BK moves")
+        if self.options.egg_behaviour.value == 1 and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have Randomize Starting Egg without randomizing moves and randomizing BK moves")
+        elif self.options.egg_behaviour.value == 2 and (self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive Eggs without randomizing moves")
+        if self.options.progressive_shoes.value == True and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive Shoes without randomizing moves and randomizing BK moves")
+        if self.options.progressive_water_training.value != 0 and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive Water Training without randomizing moves and randomizing BK moves")
+        if self.options.progressive_flight.value == True and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive flight without randomizing moves and randomizing BK moves")
+        if self.options.progressive_egg_aiming.value != 0 and (self.options.randomize_bk_moves.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive egg aiming without randomizing moves and randomizing BK moves")
+        if self.options.progressive_bash_attack.value == True and (self.options.randomize_stop_n_swap.value == False or self.options.randomize_moves == False):
+            raise ValueError("You cannot have progressive bash attack without randomizing Stop N Swap and randomizing BK moves")
+        if self.options.randomize_moves == False and self.options.jamjars_silo_costs.value != 0:
```
```suggestion
        if self.options.randomize_worlds and self.options.randomize_bk_moves.value != 0 and self.options.logic_type == 0:
            raise ValueError("Randomize Worlds and Randomize BK Moves is not compatible with Beginner Logic.")
        if not self.options.randomize_notes and self.options.randomize_worlds and self.options.randomize_bk_moves.value != 0:
            if self.multiworld.players == 1:
                raise ValueError("Randomize Notes is required for Randomize BK Moves and Randomize Worlds enabled.")
        if not self.options.randomize_notes and (self.options.extra_trebleclefs_count.value != 0 and self.options.bassclef_amount.value != 0):
            raise ValueError("Randomize Notes is required to add extra Treble Clefs or Bass Clefs")
        if self.options.progressive_beak_buster and (not self.options.randomize_bk_moves or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive Beak Buster without randomizing moves and randomizing BK moves")
        if self.options.egg_behaviour.value == 1 and (not self.options.randomize_bk_moves.value or not self.options.randomize_moves):
            raise ValueError("You cannot have Randomize Starting Egg without randomizing moves and randomizing BK moves")
        elif self.options.egg_behaviour.value == 2 and not self.options.randomize_moves:
            raise ValueError("You cannot have progressive Eggs without randomizing moves")
        if self.options.progressive_shoes and (not self.options.randomize_bk_moves or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive Shoes without randomizing moves and randomizing BK moves")
        if self.options.progressive_water_training.value != 0 and (not self.options.randomize_bk_moves or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive Water Training without randomizing moves and randomizing BK moves")
        if self.options.progressive_flight.value and (not self.options.randomize_bk_moves or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive flight without randomizing moves and randomizing BK moves")
        if self.options.progressive_egg_aiming.value != 0 and (not self.options.randomize_bk_moves or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive egg aiming without randomizing moves and randomizing BK moves")
        if self.options.progressive_bash_attack and (not self.options.randomize_stop_n_swap or not self.options.randomize_moves):
            raise ValueError("You cannot have progressive bash attack without randomizing Stop N Swap and randomizing BK moves")
        if not self.options.randomize_moves and self.options.jamjars_silo_costs.value != 0:
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+            
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        btoptions = {}
```
Would recommend looking into `self.options.as_dict()` to cut down on a majority of this.

g0goTAS replied:
I'll let jj do that

jjjj12212 replied:
thats...a big change that would need to modify both the lua end + the Connector for the Everdrive side...
May look into this over the weekend when I have more coding time.

jjjj12212 replied:
Done.

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        add_loading_zone_information(hints, HPLoc_table, get_entrance(regionName.HP))
+        add_loading_zone_information(hints, CCLoc_table, get_entrance(regionName.CC))
+        
+        hint_data.update({self.player: hints})
```
```suggestion
        hint_data.update({self.player: hints})
        
```

Silvris commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+            else:
+                return level
+
+        if self.options.randomize_world_loading_zone.value == False:
```
```suggestion
        if not self.options.randomize_world_loading_zone:
```

g0goTAS replied:
I'll let jj fix errors of that sort in the BT Client

Silvris commented on worlds/banjo_tooie/test/test_adv_move.py:
```diff
@@ -0,0 +1,163 @@
+    }
+    def test_item_pool(self) -> None:
+        super().item_pool()
+    def test_all_state_can_reach_everything(self):
+        return super().test_all_state_can_reach_everything()
+    def test_empty_state_can_reach_something(self):
+        return super().test_empty_state_can_reach_something()
+    def test_fill(self):
+        return super().test_fill()
```
Only writing this once, all three of these tests are always run on classes inheriting TestBase unless you explicitly disable them, so there's no need to override.

g0goTAS replied:
I'll let jj take care of this one

Silvris commented on worlds/banjo_tooie/test/test_treble.py:
```diff
@@ -0,0 +1,171 @@
+    def test_empty_state_can_reach_something(self):
+        return super().test_empty_state_can_reach_something()
+    def test_fill(self):
+        return super().test_fill()
```
```suggestion
        return super().test_fill()
        
```

Silvris reviewed (commented):
Mostly a quick overview for style and common pitfalls. Did not read through rules or client.

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+class EnableMultiWorldChuffyTrain(Toggle):
+    """Chuffy is lost somewhere in the MultiWorld. 
+    Once received, you can call Chuffy at any unlocked station without defeating Old King Coal."""
+    display_name = "Chuffy as a randomized AP Item"
```
```suggestion
    display_name = "Chuffy as a Randomized AP Item"
```
Other options are in title case, so this should be too

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+
+class BassclefNotes(Range):
+    """Convert some 5 notes into Bassclefs (10 notes). How many notes do you want converted?
+       Be aware that 1 bassclef removes two 5 notes and adds an additional Big-O-Pants.
```
```suggestion
       Be aware that 1 Bassclef removes two 5 notes and adds an additional Big-O-Pants.
```
Bassclef is capitalized in the previous sentence, so it should be capitalized here too

g0goTAS replied:
Bass Clef has been capitalised and split into 2 words everywhere except the connector. I'll let jj take care of the connector.

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Traps"
+
+class EnableMultiWorldDinoRoar(Toggle):
+    """Baby T-Rex Roar is lost across the MultiWorld. Other players need to help him learn to ROAR!"""
```
```suggestion
    """Baby T-Rex's Roar is lost across the MultiWorld. Other players need to help him learn to ROAR!"""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Baby T-Rex Roar"
+
+class EnableNestsanity(Toggle):
+    """Eggs and feather nests give checks when you collect them for the first time. They behave as regular egg nests the other times."""
```
```suggestion
    """Eggs and feather nests give checks when you collect them for the first time. They behave as regular egg nests after they have been collected."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Nestsanity"
+    
+class TrapsToNestRatio(Range):
+    """Select a percentage of all feather and egg nests items to be replaced with trap items
+    Requires Traps and Nestsanity to have an effect."""
```
```suggestion
    """Select a percentage of feather and egg nests items to be replaced with trap items.
    Requires Traps and Nestsanity to have an effect."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Skip Puzzles"
+
+class Backdoors(Toggle):
+    """Opens many one way switches on game start, allowing for more backdoor access to levels.
+    The following gates are preopened: MT>TDL, MT>HFP, GGM>WW, WW>TDL
+    For MT>TDL only the gate accessed from TDL's side is opened and for GGM>WW the boulders are still unexploded
+    The bridge from HFPs entrance is pre-moved to allow secondary access to clifftop
+    George is also pre-dropped to make HFP>JRL more accessible"""
```
```suggestion
    """Opens many one way switches on game start, allowing for more backdoor access to levels.
    The following gates are preopened: MT>TDL, MT>HFP, GGM>WW, WW>TDL.
    For MT>TDL only the gate accessed from TDL's side is opened and for GGM>WW the boulders are still intact.
    The bridge from HFP's entrance is pre-moved to allow secondary access to clifftop.
    George is pre-dropped to make HFP>JRL more accessible."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Open Backdoors"
+
+class OpenHag1(DefaultOnToggle):
+    """HAG 1 boss fight is opened when Cauldron Keep is opened. Only 55 jiggies are needed to win."""
```
```suggestion
    """HAG 1 boss fight is opened when Cauldron Keep is opened. Only 55 Jiggies are needed to win."""
```
Consistent capitalization for Jiggy

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "HAG 1 Open"
+
+class RandomizeWorlds(Toggle):
+    """Worlds will open in a randomized order. Randomized Moves and Puzzle Skip Required."""
```
```suggestion
    """Worlds will open in a randomized order. Randomized Moves and Skip Puzzles required."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 1
+
+class LogicType(Choice):
+    """Choose your logic difficulty and difficulty of tricks you are expected to perform to reach certain areas.
+    Please be aware that if you plan on randomizing worlds with BK Moves in the pool, you cannot use "intended" logic.""" 
```
What does the bit "you cannot use 'intended' logic" mean?

g0goTAS replied:
The option to select the logic used gives 4 options: intended, easy tricks, hard tricks, glitches. Intended logic is the name for one of them, which sticks to intended strategies only.

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 0
+
+class Silos(Choice):
+    """Choose if you want IoH Silos to be closed, randomly open 1 or enable all. If you enabled Randomized Worlds with BK Moves randomized and
+       silos set to none, it will be enforced to one."""
```
```suggestion
    """Choose if you want IoH Silos to be closed, randomly open 1, or enable all. If you enabled Randomized Worlds with BK Moves randomized and
       silos set to none, it will be enforced to one."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 0
+
+class VictoryCondition(Choice):
+    """Choose which victory condition you want
+    HAG1: Unlock the HAG1 fight and defeat Gruntilda
+    Minigame Hunt: Clear the 14 minigames and the final Canary Mary race in Cloud Cuckcoo Land to collect Mumbo Tokens
+    Boss Hunt: Kill the 8 world bosses and collect their Mumbo Tokens
+    Jinjo Family Rescue: Rescue Jinjo Families to collect their prized Mumbo Tokens
+    Wonderwing Challenge: Collect all 32 Mumbo Tokens across all boss fights, mini games and every Jinjo family
+        to gain access to HAG1 and Defeat Grunty. The Ultimate Banjo Tooie experience!!
+    Token Hunt: Mumbo's Tokens are scattered around the world. Help him find them"""
```
```suggestion
    """Choose the victory condition.
    HAG1: Unlock the HAG1 fight and defeat Gruntilda.
    Minigame Hunt: Clear the 14 minigames and the final Canary Mary race in Cloud Cuckcoo Land to collect Mumbo Tokens.
    Boss Hunt: Defeat the 8 world bosses and collect their Mumbo Tokens.
    Jinjo Family Rescue: Rescue Jinjo Families to collect their prized Mumbo Tokens.
    Wonderwing Challenge: Collect all 32 Mumbo Tokens across all boss fights, mini games, and every Jinjo family
        to gain access to HAG1 and Defeat Grunty. The Ultimate Banjo Tooie experience!!
    Token Hunt: Mumbo's Tokens are scattered around the world. Help him find them!"""
```
Would also recommend seeing how these options display on webhost in the tool tips, it can be a little weird sometimes.

jjjj12212 replied:
![image](https://github.com/user-attachments/assets/dcc2d2a5-1728-4405-86ae-f7e9c24d6de5)

ScipioWright replied:
Yeah there's that little indentation there at "to gain access to HAG1" that looks a little weird

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+
+class MinigameHuntLength(Range):
+    """How many Mumbo Tokens are needed to clear the Minigame Hunt.
+    Choose a value between 1 and 15"""
```
```suggestion
    Choose a value between 1 and 15."""
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+
+class BossHuntLength(Range):
+    """How many Mumbo Tokens are needed to clear the Boss Hunt.
+    Choose a value between 1 and 8"""
```
```suggestion
    Choose a value between 1 and 8."""
```
I'm not entirely sure it's necessary to state the range here, but I'd need to look at another yaml to confirm that.

ScipioWright reviewed (changes requested):
Lots of misc cleanup stuff mostly

g0goTAS commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 0
+
+class EnableMultiWorldBKMoveList(Choice):
+    """Banjo-Kazooie's Movelist is locked between the MultiWorld. Other players need to unlock Banjo's Moves.
```
Banjo-Kazooie refers to the title of the first game, not the state of Banjo and Kazooie being together.

ScipioWright replied:
So it's the first game's movelist that gets locked behind the multiworld?

jjjj12212 replied:
yes.

g0goTAS commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 1
+
+class CustomWorlds(FreeText):
+    """Enter a list of jiggy requirements you want for each world unlock. Max values of each world are: 1,20,30,40,50,60,70,80,90
+    Enter your custom world jiggy requirement list"""
```
It is true

g0goTAS commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+                hint_information.update({data.btid: entrance})
+
+        def get_entrance(level: str):
+            # TODO: Fix level names here too
```
I'll let jj take care of this one

jjjj12212 replied:
no longer valid. removed and cleaned up.

g0goTAS committed:
ScipioWright's initial suggestions

except 2 things

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    default = 0
+
+class EnableMultiWorldBKMoveList(Choice):
+    """Banjo-Kazooie's Movelist is locked behind  the MultiWorld.
```
```suggestion
    """Banjo-Kazooie's Movelist is locked behind the MultiWorld.
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Randomize Note Nests"
+
+class BassClefNotes(Range):
+    """Convert some 5 notes into Bass Clefs (10 notes). How many notes do you want converted?
+       Be aware that 1 bass Clef removes two 5 notes and adds an additional Big-O-Pants.
```
```suggestion
    """Convert some 5 notes into Bass Clefs (10 notes). How many notes do you want converted?
       Be aware that 1 Bass Clef removes two 5 notes and adds an additional Big-O-Pants.
```

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Traps"
+
+class EnableMultiWorldDinoRoar(Toggle):
+    """Baby T-Rex Roar's is lost across the MultiWorld. Other players need to help him learn to ROAR!"""
```
```suggestion
    """Baby T-Rex's Roar is lost across the MultiWorld. Other players need to help him learn to ROAR!"""
```

ahmed-dardery replied:
fixed

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "HAG 1 Open"
+
+class RandomizeWorlds(Toggle):
+    """Worlds will open in a randomized order. Randomized Moves and Skip Puzzle Required."""
```
```suggestion
    """Worlds will open in a randomized order. Randomized Moves and Skip Puzzles Required."""
```
The display name for the other option has an s at the end

ahmed-dardery replied:
yep

ScipioWright commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+
+class GameLength(Choice):
+    """Choose how quickly the worlds open.
+    quick: Worlds open  at 1, 3, 6, 10, 15, 21, 28, 36, and 44 Jiggys
+    normal: Worlds open  at 1, 4, 8, 14, 20, 28, 36, 45, and 55 Jiggys
+    long: Worlds open  at 1, 8, 16, 25, 34, 43, 52, 61, and 70 Jiggys
```
```suggestion
    quick: Worlds open at 1, 3, 6, 10, 15, 21, 28, 36, and 44 Jiggies
    normal: Worlds open at 1, 4, 8, 14, 20, 28, 36, 45, and 55 Jiggies
    long: Worlds open at 1, 8, 16, 25, 34, 43, 52, 61, and 70 Jiggies
```

ahmed-dardery replied:
all double spaces fixed

g0goTAS commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,874 @@
+        return all_item_table[itemname].btid
+
+    def create_item(self, itemname: str) -> Item:
+        banjoItem = all_item_table.get(itemname)
```
I'll let jj take care of this

g0goTAS commented on worlds/banjo_tooie/Rules.py:
```diff
@@ -0,0 +1,8156 @@
+# I don't know what is going on here, but it works.
+if TYPE_CHECKING:
+    from . import BanjoTooieWorld
+else:
+    BanjoTooieWorld = object
```
I'll let jj take care of this

jjjj12212 replied:
If I remove this, I get the following error:
Please fix your YAML.
 No Functional world found to handle game Banjo-Tooie. did you mean 'banjo_tooie'?

So we still need this for some reason.

jjjj12212 replied:
nvm, all is good. it generates fine. User error :P

Commit:
trailing whitespaces and adding new lines to end of files

g0goTAS committed:
Silvris suggestions: bools and enums

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

Commit:
fix broken tests and remove redundant overrides

Commit:
removing double spaces

ahmed-dardery commented on worlds/banjo_tooie/Names/itemName.py:
```diff
@@ -0,0 +1,141 @@
+
+GNEST = "Golden Egg Nest"
+ENEST = "Egg Nest"
+FNEST = "Feather Nest"
```
All files now have new lines at the end and no trailing spaces.

g0goTAS committed:
bools and enums: the sequel

Commit:
removed one more space than necessary xD

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

g0goTAS committed:
fix auto-completion that failed

g0goTAS committed:
wing whack doubloons can now have moves on them

jjjj12212 committed:
Remove todo and remove if branch that the code will never go into.

jjjj12212 committed:
Include French guide that G0go made

NewSoupVi commented on worlds/banjo_tooie/Items.py:
```diff
@@ -0,0 +1,211 @@
+}
+
+all_group_table = {
+    'jiggy': jiggy_table,
```
To specify a bit further, please use double-quotes for strings specifically.

https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/style.md

ScipioWright commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -576,20 +576,14 @@ def pre_fill_me(self) -> None:
-                    self.get_location("World "+ str(world_num) +" Unlocked").place_locked_item(item)
-                    world_num = world_num + 1
+                self.get_location("World "+ str(world_num) +" Unlocked").place_locked_item(item)
+                world_num = world_num + 1
```
For the future, you can do `world_num += 1` and it'll have the same effect o: No real need to change it, just a convenience for future stuff

ScipioWright commented:
Would also recommend hitting the Resolve button on comments where you've either fixed the thing or declined to fix it o:

jjjj12212 committed:
.value after victory_condition

jjjj12212 committed:
import cleanup

jjjj12212 committed:
convert majority of single quotes to double quotes

jjjj12212 commented on worlds/banjo_tooie/Options.py:
```diff
@@ -0,0 +1,513 @@
+    display_name = "Cheato Rewards"
+
+class EnableMultiWorldJinjos(DefaultOnToggle):
+    """Jinjos have fled to other worlds. Other players need to return them home."""
```
Will do!

jjjj12212 committed:
changes in Options descriptions

jjjj12212 committed:
fix wording in doc.

jjjj12212 committed:
another option wording change + recommended change.

Commit:
cleanup of all options

g0goTAS commented:
I read the comment about the resolve button, but the button does not seem to show for me, and I'm not sure why. Is it because someone put me as a reviewer, or perhaps because I'm not the person that created the PR?

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
remove unused variable

jjjj12212 committed:
Slot Changes

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
additional PR changes

jjjj12212 committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

jjjj12212 committed:
finish up this request.

jjjj12212 commented:
Discovered a couple of bugs while converting the Slots... Working out the kinks.

jjjj12212 commented:
bugs fixed, just wait to test the latest features with the community before we push the update.

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
Banjo-Tooie V4.2 which includes Signposts hints and fixes

jjjj12212 commented:
We have pushed the latest changes.
This includes Signpost hints and more bug fixes.

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
V4.4.2

g0goTAS commented:
Due to the fact that we continue on adding features while this pull requests is on hold, we would like to get a heads up when the process of review continues. We wouldn't want the version that's on AP to be missing important features.

ScipioWright commented on worlds/banjo_tooie/docs/en_Banjo-Tooie.md:
```diff
@@ -0,0 +1,56 @@
+- Prison Code Door is always open
+- Allows for instant transformation once Mumbo Skull or Humba Wigwam have been tagged by Banjo & Kazooie
+- Option to customize the dialog character when receiving certain items
+- Pause menu contains a new Arhipelago Menu
```
```suggestion
- Pause menu contains a new Archipelago Menu
```

duckboycool commented on worlds/banjo_tooie/test/test_victory.py:
```diff
@@ -0,0 +1,534 @@
+#Tests and make sure that if the correct Victory Condition is set, enough Mumbo Tokens are placed
+# and the game is beatable.
+class TokenTest(BanjoTooieTestBase):
+    mumbo_token_location_group = {}
```
```suggestion
    mumbo_token_location_group = set()
```
It looks like this was meant to be a set while `{}` is an empty `dict`.

duckboycool reviewed (changes requested):
Not a full review, I looked at typing errors and found a few thing including a couple mistakes that affect functionality. Also for some reason the review comments later on in `Rules.py` seem to be broken on my end, so the ones about not calling a function are supposed to be in `notes_cheese` and `notes_sack_race` if they don't show up that way.

g0goTAS commented on worlds/banjo_tooie/Rules.py:
```diff
@@ -0,0 +1,9084 @@
+            itemName.SAPACK,
+        ]
+
+        self.magic_forbid = {
```
They were used in very early versions of the rando. They will be removed once jj updates the PR.

g0goTAS commented:
Allright, all of those have been fixed for the next update. I'll let jj update the PR once he's ready to do it.

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
V4.8 of the latest Banjo-Tooie AP

ScipioWright committed:
Fix up hints.py

ScipioWright commented on worlds/banjo_tooie/WorldOrder.py:
```diff
@@ -0,0 +1,433 @@
+
+    return worlds
+
+def set_level_costs(world: BanjoTooieWorld) -> None:
+    normal_costs = [1,4,8,14,20,28,36,45,55]
+    quick_costs = [1,3,6,10,15,21,28,35,44]
+    long_costs = [1,8,16,25,34,43,52,60,70]
+    level_cost_max = [1,10,20,30,50,60,70,80,90]
+    custom_costs = world.options.custom_worlds.value.split(",")
+
+    random_costs = [1,1,1,1,1,1,1,1,1]
+    for i in range(len(random_costs)):
+        random_costs[i] = world.random.randint(1, level_cost_max[i])
+
+    if len(custom_costs) < 9:
+        raise ValueError("Custom Costs has too few levels.")
+
+    if len(custom_costs) > 9:
+        raise ValueError("Custom Costs has too many levels.")
+
+    for i in range(len(custom_costs)):
+        try:
+            int(custom_costs[i])
+        except Exception:
+            raise ValueError("Custom Cost for world "+str(i+1)+" must be a number.")
+        custom_costs[i] = int(custom_costs[i])
+        if custom_costs[i] == 0:
+            raise ValueError("Custom Cost for world "+str(i+1)+" can't be 0.")
+        if custom_costs[i] < 0:
+            raise ValueError("Custom Cost for world "+str(i+1)+" can't be negative.")
+        if custom_costs[i] > level_cost_max[i]:
+            raise ValueError("Custom Cost for world "+str(i+1)+" is too high.")
```
```suggestion

def set_level_costs(world: BanjoTooieWorld) -> None:
    normal_costs = [1, 4, 8, 14, 20, 28, 36, 45, 55]
    quick_costs = [1, 3, 6, 10, 15, 21, 28, 35, 44]
    long_costs = [1, 8, 16, 25, 34, 43, 52, 60, 70]
    level_cost_max = [1, 10, 20, 30, 50, 60, 70, 80, 90]
    try:
        custom_costs = [int(cost) for cost in world.options.custom_worlds.value.split(",")]
    except ValueError:
        raise OptionError(f"Custom Costs for {world.player_name} must be numeric")

    random_costs = [1, 1, 1, 1, 1, 1, 1, 1, 1]
    for i in range(len(random_costs)):
        random_costs[i] = world.random.randint(1, level_cost_max[i])

    if len(custom_costs) < 9:
        raise ValueError("Custom Costs has too few levels.")

    if len(custom_costs) > 9:
        raise ValueError("Custom Costs has too many levels.")

    for i in range(len(custom_costs)):
        if custom_costs[i] == 0:
            raise ValueError(f"Custom Cost for world {i + 1} can't be 0.")
        if custom_costs[i] < 0:
            raise ValueError(f"Custom Cost for world {i + 1} can't be negative.")
        if custom_costs[i] > level_cost_max[i]:
            raise ValueError(f"Custom Cost for world {i + 1} is too high.")
```

Moved the try except earlier so that the typing was nicer
Made the value errors use f strings instead of concatenation since it's easier to read (and it just lets you use ints)

ScipioWright commented on worlds/banjo_tooie/Names/regionName.py:
```diff
@@ -0,0 +1,128 @@
+CHUFFY = "Chuffy's Cabin"
+
+# These regions represent the act of using warps between different sections of a world.
+IHSILOS = "Isle O' Hag Silos"
+MTWARP = "Mayahem Temple Warp Pads"
+GMWARP = "Glitter Gulch Mine Warp Pads"
+WWWARP = "Witchyworld Warp Pads"
+JRWARP = "Jolly Roger Lagoon Warp Pads"
+TLWARP = "Terrydactyland Warp Pads"
+GIWARP = "Grunty Industries Warp Pads"
+HPWARP = "Hailfire Peaks Warp Pads"
+CCWARP = "Cloud Cuckooland Warp Pads"
+CKWARP = "Cauldron Keep Warp Pads"
```
`MTWARP`, `WWWARP`, `JRWARP`, `TLWARP`, and `GIWARP` are defined twice

ScipioWright reviewed (commented):
Bunch of minor stuff, need more time to review some more stuff since there's a lot here

g0goTAS committed:
Fixed Settings class indentation

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Removed note about older Python versions

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Fixed guide linting

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Removed excess type declaration

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Prettier location groups

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
added missing space

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Removed unused variable

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
added missing space

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Prettier custom world costs

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Removed warp pad region duplicates

g0goTAS committed:
Fix up hints.py

style fixes

ScipioWright reviewed (approved):
Test gens worked out fine, code review went fine. Some nitpicks but that's not my job!
Did not test the client.
Did not verify logic.

g0goTAS committed:
Merge branch 'main' into main

g0goTAS committed:
Update to Tooie 4.11.3

Main changes:
* Cryptic hint overhaul
* yamlless generation for UT
* AP0.6.3 features
* archipelago.json
* fixed rare generation fails
* logic fixes
* fixed a lot of linting errors (some of which were already fixed in this PR)

Full changelog: https://github.com/jjjj12212/Archipelago-BanjoTooie/compare/V4.8.1...V4.11.3#diff-c275fec9c453c2a42515bc5ab47e30fa4130ff1426aef2f6e000a9a34e122cb8

g0goTAS commented:
Since the verification process has gotten stale with this PR, the PR has been updated to use the latest version of the Tooie rando.

g0goTAS commented:
Currently, the Banjo-Tooie apworld logs the time for hint generation manually (https://github.com/ArchipelagoMW/Archipelago/pull/4471/files#diff-864f12b4cf36b500b8adfb0e5ca1d44625298764b89f13763e5c89df519e3583R1007). Depending on what happens with https://github.com/ArchipelagoMW/Archipelago/pull/5663, this may need to be removed.

duckboycool commented on worlds/banjo_tooie/test/confidence.py:
```diff
@@ -0,0 +1,49 @@
+    Don't use this directly in tests, instead, you can run it locally to make sure the generated numbers
+    make sense.
+
+
```
Extra newlines.

g0goTAS replied:
fixed

duckboycool commented on worlds/banjo_tooie/test/test_jinjo.py:
```diff
@@ -0,0 +1,266 @@
+                        placed_correctly += 1
+                    location_item = self.multiworld.get_location(locationName.JINJOGI3, self.player).item.name
+                    if location_item == name:
+                        placed_correctly += 1
```
This could be turned into a for loop over a list of the location names.

duckboycool commented on worlds/banjo_tooie/test/test_logic.py:
```diff
@@ -0,0 +1,68 @@
+        "randomize_bt_moves": RandomizeBTMoveList.option_true,
+        "randomize_bk_moves": RandomizeBKMoveList.option_all,
+        "randomize_notes": RandomizeNotes.option_true
+
```
```suggestion
```
Extra newline.

duckboycool commented on worlds/banjo_tooie/test/test_notes.py:
```diff
@@ -0,0 +1,106 @@
+
+        progresssion = sum(1 for item in notes_in_pool if item.advancement)
+        useful = sum(1 for item in notes_in_pool if item.useful)
+        filler = sum(1 for item in notes_in_pool if item.filler)
```
```suggestion
```
This var's just unused.

duckboycool commented on worlds/banjo_tooie/test/test_notes.py:
```diff
@@ -0,0 +1,106 @@
+        assert item_pool_names.count(itemName.NOTE) == 0
+
+    def test_prefills(self) -> None:
+        vanilla_locations_names = [location_name for location_name, location_data in all_location_table.items()
```
Could this just use `location_name_groups`?

duckboycool commented on worlds/banjo_tooie/test/test_signpost_hints.py:
```diff
@@ -0,0 +1,191 @@
+                and location.player == hint_data.location_player_id
+            ][0]
+
+            text = hint_data.text[::]
```
Is there a reason to not just use `hint_data.text` below (or remove the copy through slicing here)?

duckboycool commented on worlds/banjo_tooie/test/test_train.py:
```diff
@@ -0,0 +1,195 @@
+from worlds.banjo_tooie.Names import locationName
+from ..Names import itemName
```
Another combination of imports here.

g0goTAS replied:
Fixed.

duckboycool commented on worlds/banjo_tooie/test/test_victory.py:
```diff
@@ -0,0 +1,606 @@
+from .test_logic import EasyTricksLogic, GlitchesLogic, HardTricksLogic, IntendedLogic
+from . import BanjoTooieTestBase
+from ..Names import locationName, itemName, regionName
+from .. import all_item_table, all_group_table
```
```suggestion
from ..Names import locationName
```
Unused imports.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+    IHPLLoc_table, IHPGLoc_table, IHCTLoc_table, IHWLLoc_table, IHQMLoc_table, \
+    CheatoRewardsLoc_table, JinjoRewardsLoc_table, HoneyBRewardsLoc_table
+from .Regions import create_regions, connect_regions
+from .Options import BanjoTooieOptions, EggsBehaviour, HintClarity, JamjarsSiloCosts, LogicType, ProgressiveEggAim, \
```
```suggestion
from .Options import BanjoTooieOptions, EggsBehaviour, JamjarsSiloCosts, LogicType, ProgressiveEggAim, \
```
Unused import.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+    glitches_item_name = itemName.UT_GLITCHED
+    ut_can_gen_without_yaml = True
+
+    location_name_to_id = {name: data.btid for name, data in all_location_table.items()}
```
It might be best to not include events in here in case the automatic filter for that is removed, so if it doesn't actually need to be in `all_location_table`, you could remove what I guess is the singular event from there.

g0goTAS replied:
Added a filter for entries that have a None id.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+        return created_item
+
+    def calculate_useful_filler(self, total, progression) -> tuple[int, int]:
+        # We want to split non-progressive items into two halfs
```
I assume this comment means "non-progression" items, although I'm somewhat confused about the purpose of doing this. Is there a reason why some but not all of these items should be excludable?

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+            filler_notes += useful_notes
+            useful_notes = 0
+        if filler_notes < 0:
+            warnings.warn("Number of notes that need to be inserted is somehow negative.")
```
This should probably just log a warning/error instead of raising one with `warnings`.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+    def choose_starter_attack(self) -> None:
+        if self.options.randomize_bk_moves != RandomizeBKMoveList.option_none:
+            chosen_attack: str
+            base_attacks: list
```
```suggestion
            chosen_attack: str
            base_attacks: list
```
These annotations aren't really necessary, and typing just as `list` without saying the contained type actually prevents it from being inferred in the assignments below.

duckboycool replied:
Annotations don't actually do a declaration in python, they're just to show types of values (unless it's something like dataclasses that look at the annotations and generate values from them). if/else blocks don't make their own scope, so they'd be visible after.

g0goTAS replied:
Allright, removed both of them.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+
+            item = self.create_item(itemName.OJINJO)
+            self.get_location(locationName.JINJOWW4).place_locked_item(item)
+            self.get_location(locationName.JINJOHP2).place_locked_item(item)
```
```suggestion
            locations = [locationName.JINJOWW4, locationName.JINJOHP2]
            for loc in locations:
                self.get_location(loc).place_locked_item(self.create_item(itemName.OJINJO))
```
Not a big deal for the two here, but I feel like it'd be better for these to be replaced by a loop similar to this. Also I'm not sure if matters for locked items, but I feel like it's best to avoid re-using the same item instance and instead create a new one like for Jiggys here.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+            self.get_location(locationName.JINJOGI3).place_locked_item(item)
+
+    def allow_jiggies_as_filler(self) -> bool:
+        return self.options.replace_extra_jiggies and self.jiggies_in_pool < self.hard_item_limit
```
```suggestion
        return bool(self.options.replace_extra_jiggies) and self.jiggies_in_pool < self.hard_item_limit
```
or
```suggestion
        return bool(self.options.replace_extra_jiggies and self.jiggies_in_pool < self.hard_item_limit)
```
Don't think it should really matter with how they're used, but technically using the options like this in logical expressions means they won't return a `bool` like the function def says.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+            regionName.CCBOSS: regionName.CC,
+        }
+        bt_players = world.get_game_players(cls.game)
+        spoiler_handle.write('\n\nBanjo-Tooie ({})'.format(BanjoTooieWorld.version))
```
```suggestion
        spoiler_handle.write(f"\n\nBanjo-Tooie (V{cls.world_version.as_simple_string()})")
```
Why not use an f-string here? Also here's an example of how you can use `world_version`.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -0,0 +1,1094 @@
+
+        # Elements that are randomised outside the yaml and affects gameplay
+        custom_bt_data: Dict[str, Any] = {
+            "player_name": self.multiworld.player_name[self.player],
```
```suggestion
            "player_name": self.player_name,
```
There's a helper for this now.

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+script_version: int = 5
+version: str = "V4.11.3"
+patch_md5: str = "54efb965e58f1bb15487d411dc8955a3"
+bt_options = settings.get_settings().banjo_tooie_options
```
It might be better to import the world to get this data, e.g. `BanjoTooieWorld.settings` or `BanjoTooieWorld.location_name_to_id` instead of `get_settings()` or `network_data_package` (which means importing *all* worlds). Some other worlds' clients seem to this.

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+  with open(path, "wb") as fi:
+    fi.write(data)
+
+def open_world_file(resource: str, mode: str = "rb", encoding: str = None):
```
Could this be replaced by `pkgutil.get_data`?

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+    def on_print_json(self, args: dict):
+        if self.ui:
+            self.ui.print_json(copy.deepcopy(args["data"]))
+            relevant = args.get("type", None) in {"ItemSend"}
```
```suggestion
            relevant = args.get("type") == "ItemSend"
```
Slightly overcomplicated.

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+                    relevant = True
+
+                if relevant == True:
+                    msg = self.raw_text_parser(copy.deepcopy(args["data"]))
```
The place where this is used is commented out. Also would it actually be modified to make the `copy.deepcopy` needed anyway?

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,578 @@
+        REQUIRED_BY_MULTIWORLD = 1,
+        NOT_REQUIRED = 2
+
+    item_requirement_cache: dict[Location, "Hint.ItemRequirement"] = dict()
```
```suggestion
    item_requirement_cache: dict[Location, ItemRequirement] = {}
```
Quoted type not needed here.

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,578 @@
+
+                Hint.item_requirement_cache[hinted_location] = item_required
+
+    def is_last_cryptic_hint_world(world: "BanjoTooieWorld"):
```
```suggestion
    @staticmethod
    def is_last_cryptic_hint_world(world: "BanjoTooieWorld"):
```
Same here.

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,986 @@
+        return
+
+    def _set_message(self, msg: dict, msg_id: Union[int, None]):
+        if msg_id == None:
```
Is there a reason this was marked as resolved without changes?

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,578 @@
+
+
+def get_location_by_name(world: "BanjoTooieWorld", name: str) -> Location | None:
+    potential_match = list(filter(lambda location: location.name == name, get_player_hintable_locations(world)))
```
```suggestion
    potential_match = [location for location in get_player_hintable_locations(world) if location.name == name]
```
Don't see why not to use a comprehension for this one.

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,578 @@
+
+
+def get_player_hintable_locations(world: "BanjoTooieWorld") -> List[Location]:
+    return [location for location in world.multiworld.get_locations(world.player) if should_consider_location(location)]
```
```suggestion
    return [location for location in world.get_locations() if should_consider_location(location)]
```
Helper for this.

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,578 @@
+
+def get_signpost_location_ids() -> List[int]:
+    location_datas = list(filter(lambda location_data: location_data.group == "Signpost", all_location_table.values()))
+    return [location_data.btid for location_data in location_datas]
```
```suggestion
    return [location_data.btid for location_data in all_location_table.values() if location_data.group == "Signpost"]
```
This one can also be condensed into a single expression.

duckboycool commented on worlds/banjo_tooie/Items.py:
```diff
@@ -0,0 +1,298 @@
+    btid: int | None = None
+    qty: int = 0
+    type: ItemClassification = ItemClassification.progression
+    default_location: str = ""
```
These are always set anyway, so they don't really need to be there as defaults. Also it might be better to make `default_location` None when there isn't one as opposed to an empty string.

duckboycool commented on worlds/banjo_tooie/Regions.py:
```diff
@@ -0,0 +1,2243 @@
+
+
+def create_region(multiworld, player: int, active_locations, name: str, locations=None):
+    ret = Region(name, player, multiworld)
```
```suggestion
def create_region(world, active_locations, name: str, locations=None):
    ret = Region(name, world.player, world.multiworld)
```
Seems like it would probably make more sense to pass world here and then use that instead of `multiworld.worlds[player]`. Also it seems like the if chain could be replaced by an `options.victory_condition in` similar to above.

duckboycool reviewed (changes requested):
I have reviewed the world, test, and most of the client code, but I have not done any functional testing past unittests, reviewed logic, looked at connector code, or looked at `WorldOrder.py` since I'm not really sure what it does.

Some general comments:
- There are a really high amount of test combinations (a lot of them with duplicated code as well). I had around 269,000 total subtests running the world tests, and it took nearly 7 minutes single-core. (For reference, Stardew has around 20k total subtests and it runs in under 40 seconds.) Even if you do want to keep them all, it's probably not very valuable for most of these tests combinations to have `run_default_tests`. There are also several world tests that fail currently for me, in addition to some core tests.
- Most places in the code use old-style typing, which is technically deprecated, although a removal version has not been decided yet. It may be easier to remove now compared to later if it does end up being removed.

duckboycool reviewed (changes requested):
I guess it posted the main review comment twice when I refreshed after it failing the first time.

g0goTAS commented on worlds/banjo_tooie/Names/itemName.py:
May I ask where I can find that test?

duckboycool replied:
It's `test/general/test_packages.py`. It also causes the build task to fail (which is why there's a test).

duckboycool commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+    def __init__(self, server_address, password):
+        super().__init__(server_address, password)
+        self.game = "Banjo-Tooie"
+        self.n64_streams: (StreamReader, StreamWriter) = None # type: ignore
```
```suggestion
        self.n64_streams: tuple[StreamReader, StreamWriter] | None = None
```
I guess you're aware by the type comment, but this annotation (although it's also in other worlds) is invalid. This is how you should write it. Also you should import `StreamReader` and `StreamWriter` for this.

duckboycool reviewed (commented):
While looking at the unittests action on this fork (https://github.com/jjjj12212/AP-Tooie-PR/actions/runs/19182453556), I noticed that the analyze modified files action also failed (one reason being that `test_boss_loading zones.py`'s name having a space breaks it I guess, and also undefined vars I didn't notice before). I'd really recommend that you look at these action failures, ~~and it'd also probably be good to enable them for your development fork as well~~ the dev one isn't actually a fork, but you could still run at least unittests locally as well, or just sync onto the fork more often.

g0goTAS commented on worlds/banjo_tooie/docs/setup_en.md:
```diff
@@ -0,0 +1,70 @@
+
+# SETUP GUIDE FOR BANJO-TOOIE ARCHIPELAGO
```
Fixed.

g0goTAS committed:
periods and caps title

g0goTAS committed:
Markdown instead of tags

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
merge conflict fix

g0goTAS committed:
French setup fix

g0goTAS commented on worlds/banjo_tooie/docs/setup_fr.md:
fixed, but worth double-checking

g0goTAS commented on worlds/banjo_tooie/test/test_cheato_rewards.py:
```diff
@@ -0,0 +1,96 @@
+from BaseClasses import ItemClassification
+from worlds.banjo_tooie.Names import locationName
+from ..Names import itemName
```
Yeah, not sure what happened with the automatic imports there. Made them use the relative path.

g0goTAS committed:
use whole division

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS commented on worlds/banjo_tooie/test/test_glowbos.py:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    def test_item_pool(self) -> None:
+        item_pool_names = [item.name for item in self.multiworld.itempool]
```
Good point. Fixed.

g0goTAS commented on worlds/banjo_tooie/test/test_notes.py:
```diff
@@ -0,0 +1,106 @@
+
+    def test_item_pool(self) -> None:
+        # max jamjars cost is 765. There are 9 trebleclefs by default.
+        progression_notes_default = (765 - 9*20) / 5
```
Yes, fixed.

g0goTAS committed:
test_notes: removed unused variable

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
fixed type definition

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
Removed unused import

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
removed unused id

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
unused import

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
removed newline

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
classvar

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
valueerror

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
removed useless parenthese

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
removed extra spaces

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
list is indeed made out of list

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
F string

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
player_name

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS commented on worlds/banjo_tooie/BTClient.py:
```diff
@@ -0,0 +1,1248 @@
+CONNECTION_INITIAL_STATUS = "Connection has not been initiated"
+
+"""
+Payload: lua -> client
```
Replaced double quotes by single quotes

g0goTAS committed:
unused import

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
removed quoted type

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
staticmethod

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
staticmethod

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
simplifying the filter

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
helper

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
simple life? simple life.

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
why filter when you can list comprehension

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

g0goTAS committed:
PR multifix

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP-Tooie-PR

g0goTAS committed:
remove colon

g0goTAS commented:
Allright, I've fixed most of the things that got mentionned by duckboycool. I'll probably continue next week end. In the mean time, I'll let jj fix the BTClient and resolve the threads that I fixed (only he can resolve threads, for some reasons).

jjjj12212 committed:
Merge branch 'main' into main

g0goTAS committed:
IntFlag simplification

Co-authored-by: Duck <31627079+duckboycool@users.noreply.github.com>

duckboycool commented:
> Not sure what this comment is referring to. I can't reply to it. [#4471 (comment)](https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2558078908)

That's a reply to a previous comment, clicking that link itself actually seems like it should take you to the original. Same for this one. https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2558145507

duckboycool commented on worlds/banjo_tooie/Rules.py:
Ah, I see that these logic checks also test for the Universal Tracker Glitch placeholder and so technically will depend on state. I'm not sure how exactly Universal Tracker intends you to do this, but I assume this is an item that won't affect actual generation at all, so there should still be some way to make it work without including all these checks inside rules.

duckboycool replied:
I think that exactly one of the conditions should be met, so yes I don't think there should actually be an issue with the var being unbound.

In terms of performance, I can't say how much of this is due to this in particular, but the world appears to be slower than any of the current ones in core with the default template. OoT is known to be a slow world, but for me a test gen with 7x its template ran in just over 5 seconds (with `--skip_output`) whereas this world took 42 seconds (also the current commit on this PR is broken, I had to go back a bit). I originally tried with 15x, but it couldn't finish since the world starts taking drastically more time as I include more of it, pointing to some sort of scaling issue that I'm not sure of the cause of (it's probably not about this rules thing itself). 
Either way performance definitely does matter once the world's being included in bigger multiworlds, and I don't think it would be that difficult to edit these rules to remove the checks. Even just deleting the branches that all end up doing the same thing but call the option check up to 4 times (and consolidating branches where 2 or 3 of them do the same thing) would probably help decently.

g0goTAS replied:
I just did a test gen, just to reproduce what you are saying, and I feel like it may be more of a problem with the lack of filler items more than optimisation itself. OoT's default options probably have around 30% to 50% of progression items in its default options, which makes for an easy filling. On the other side, Tooie has around 90% of its items that are progression, which we have learned from experimentation that it makes for a really, really difficult fill with the way that AP generates seeds.

To show this difference in practice, I would invite you to try the following:
Do a control generation with however many yamls you desire of the default options. After that, do a generation with the same number of yaml, but with the "nestsanity" options turned on in all the yaml. This option doubles the number of checks in the game, and adds nothing but fillers in the pool. What you'll notice is that the second generation will finish much sooner than the first one, due to the fact that the generator has a much easier time placing items, once the concentration of progression drops from 90% to 40%, despite having double the number of checks to fill.

g0goTAS committed:
deleting useless declarations

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP-Tooie-PR

g0goTAS committed:
Removed all lambdas

g0goTAS committed:
Removed unused self

g0goTAS committed:
Fixed failing tests

g0goTAS committed:
added magic file to fix magic test

g0goTAS committed:
cryptic hint optimisation

duckboycool commented:
Is this PR branch up to date? It still fails unittests currently, and some comments are marked as fixed but haven't been changed here like https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2557393320.

g0goTAS committed:
A change of value

g0goTAS committed:
Removed unnecessary definition

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
game name not defined

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
Removed useless version

g0goTAS committed:
create_region refactor

g0goTAS committed:
fill hook magic

g0goTAS committed:
merge conflict

Auztin committed:
get data from BanjoTooieWorld

Auztin committed:
use __loader__ to load apworld files

Auztin committed:
use Counter instead

Auztin committed:
fix type annotation

Auztin committed:
remove unused function arg and fix type annotation

Auztin committed:
simplify code

Auztin committed:
remove unused variable

Auztin committed:
replace with commented lines to avoid docstring

Auztin committed:
remove useless function call

Auztin committed:
organize imports. type annotations. type fixing

Auztin committed:
use pkgutil instead

Auztin committed:
verify we're a package and that the resource was found

Auztin committed:
kivy can't be loaded early

g0goTAS committed:
Warp pad regions only exist when they're randomised

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP-Tooie-PR

jjjj12212 committed:
Merge branch 'main' into main

g0goTAS committed:
is not none optimisation

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
useful_filler fix

Co-authored-by: Scipio Wright <scipiowright@gmail.com>

g0goTAS committed:
diet test suite

g0goTAS committed:
test fix

g0goTAS committed:
logic fixes

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -1,6 +1,9 @@
 from math import ceil
 import time
+
+import loguru
```
You should add a `requirements.txt` for this new external dependency instead of relying on it already being installed for Starcraft 2, although is it really needed here compared to the regular logger already used elsewhere?

duckboycool replied:
I think just `logging.warning` (which is already imported just above) should be fine for the use you had.

duckboycool commented on worlds/banjo_tooie/Hints.py:
```diff
@@ -0,0 +1,546 @@
+import re
+from typing import Iterable, List, Set, Union, TYPE_CHECKING
+from dataclasses import dataclass
+from BaseClasses import ItemClassification, Location, LocationProgressType, CollectionState, MultiWorld
```
```suggestion
from typing import Iterable, List, Union, TYPE_CHECKING
from dataclasses import dataclass
from BaseClasses import ItemClassification, Location, LocationProgressType, CollectionState
```
The unused imports refuse to die.

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -589,114 +585,112 @@ def generate_early(self) -> None:
+                    or not self.options.randomize_bt_moves.value
                 ):
             raise OptionError("You cannot have progressive Water Training\
                 without randomizing moves and randomizing BK moves")
```
```suggestion
            raise OptionError("You cannot have progressive Water Training "
                              "without randomizing moves and randomizing BK moves")
```
Same thing as this comment. https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2557928479

duckboycool reviewed (changes requested):
I have looked over the changes made for my previous comments, and found a few things there. I think I'll be ready to approve once these are resolved.

In terms of the rules, I know that there's a lot and it would take a lot to refactor them not to contain option checks in them (although it does kinda hurt for the fake UT item to be checked a ton during real generation). But I would really like to at least see these completely redundant branches removed. Not only should this save some time in option checks, it would also just remove a ton of unnecessary repeated code here.

I'm willing to pass this fill slot data comment onto core review (https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2578968809), but I will warn you that the specific issue of including unneeded things in slot data has been a complaint of core in the past.

Also this is up to you, but the world appears to pretty much always `OptionError`s on random options, which sometimes people use for a themed game. If you want to support full-random, it might be best to make some options take priority if incompatible ones are selected.

duckboycool commented on worlds/banjo_tooie/Rules.py:
```diff
@@ -0,0 +1,10223 @@
+        elif self.easy_tricks_logic(state):
+            logic = self.long_jump(state)\
+                    or self.springy_step_shoes(state)\
+                 
```
And annoyingly, GitHub still handles the display here badly too. It's on lines 4851 to 4865.

g0goTAS committed:
Removed b tag

g0goTAS commented on worlds/banjo_tooie/docs/setup_en.md:
Fixed.

g0goTAS committed:
Missing state

g0goTAS committed:
Ending the type casting wizardry

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -89,25 +91,18 @@ class BanjoTooieWorld(World):
+    location_name_to_id = {name: data.btid for name, data in all_location_table.items() if data.btid is not None}
+    location_name_to_group = {name: data.group for name, data in all_location_table.items() if data.group is not None}
 
     item_name_groups = {
```
If you iterate over a dict, it'll be by the keys by default. Hypothetically if something tried to use some other set-specific operation it would break there though.

g0goTAS committed:
Vanilla jinjo jiggies only when vanilla jinjos

g0goTAS committed:
Wrong return type

g0goTAS commented on worlds/banjo_tooie/Rules.py:
```diff
@@ -0,0 +1,10223 @@
+            logic = self.has_green_relics(state, 10)
+        elif self.glitches_logic(state):
+            logic = self.has_green_relics(state, 10)
+        return logic
```
I see the concept, but we do not want to analyse and rewrite 10000 lines of code for a minor performance boost. That will be saved for when we'll decide to completely rewrite the logic to support a loading zone randomiser, when we're motivated enough to do that.

duckboycool committed:
Formatting changes

g0goTAS committed:
loguru.logger more like logging

duckboycool committed:
Condense rule branches

duckboycool commented:
I've opened one for it at https://github.com/jjjj12212/AP-Tooie-PR/pull/2.

duckboycool committed:
One small change

g0goTAS committed:
Merge pull request #2 from duckboycool/tooie_rules

Rules.py Refactor

duckboycool reviewed (approved):
I have not tested playing an actual world, but everything I've looked at in the world seems to be alright now. The included subset of tests is still fairly large, but reasonable as it's quicker than Stardew. Using many slots of the default options using the more restrictive default logic preset does still start to struggle, but the `fill_hook` adjustment does seem to have helped a good bit.

The world is stable on default options, and random ones with the fuzzer (after fixing `custom_worlds` to default so at least a small amount would generate without `OptionError`) also only gave a few `FillError`s.

jjjj12212 committed:
Merge branch 'main' into main

g0goTAS committed:
4.11.4

jjjj12212 committed:
Adjust OptionError Text +  BK Moves Generation Restriction

duckboycool commented on worlds/banjo_tooie/__init__.py:
```diff
@@ -588,76 +588,78 @@ def validate_yaml_options(self) -> None:
+                raise OptionError(
+                    "Banjo-Tooie: Your options are too restrictive for Randomize BK Moves. \
+                    Try randomizing notes, signs or nestsanity"
+                )
```
```suggestion
            raise OptionError(
                "Banjo-Tooie: Your options are too restrictive for Randomize BK Moves. "
                "Try randomizing notes, signs or nestsanity"
            )
```
This adds another place where there would be extra spaces included in the string from how it's written (which by the way, the last one seems to not have actually been applied here: https://github.com/ArchipelagoMW/Archipelago/pull/4471#discussion_r2644804841). Also if the if is removed, it'd be good to dedent this.

Although I'm not really sure that adding this to the messages is itself really necessary. Errors will already show which yaml they game from in the exception (unlike warnings), and I haven't seen other worlds do this generally.

jjjj12212 committed:
Adjust strings in errors

jjjj12212 committed:
fix indentation

g0goTAS committed:
Bugfixes

jjjj12212 committed:
Merge branch 'main' into main

g0goTAS committed:
indent refix

g0goTAS committed:
newline

Mysteryem committed:
Improve generation performance by removing UT logic from normal generation

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
version change

jjjj12212 committed:
Cryptic Hint Optimizations

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
V4.11.5 update

Mysteryem commented:
Eije put together a [meta file for Banjo Tooie](https://github.com/Eijebong/Archipelago-index/blob/main/fuzz-meta/banjo_tooie.yaml) for his [fuzzer](https://github.com/Eijebong/Archipelago-fuzzer) (`-m <path to meta file>`) so that Banjo Tooie can be fuzzed without constantly running into `OptionError`s. Through this, a number of issues have been identified. There are accessibility failures, missing indirect conditions, and a few seeds that put more items into the item pool than have unfilled locations. I have compiled these failures into a zip, along with some general FillErrors.

All fuzzer generations I did were with a single Banjo Tooie slot:

[fuzz_output_banjo_tooie.zip](https://github.com/user-attachments/files/26871660/fuzz_output_banjo_tooie.zip)

For general diagnosis of failing seeds, each fuzzer failure provides the generation log (where you can get the seed), and the randomly generated yaml that was used. The yaml can be placed in the Players folder and then generation with the same seed can be run with a debugger (`--seed <seed here>` for specifying a seed to `Generate.py`).

### Accessibility Failures
In `full` accessibility, this often means that some locations are not reachable even with every item in the pool/pre-placed.

In `minimal` accessibility, I think this only means that the multiworld was not beatable (the spheres iteration ran out of reachable locations without beating the goal).

I checked the accessibility failure in generation `142` in a debugger (a `minimal` accessibility failure), it seems this multiworld is unbeatable both immediately before and after the `pre_fill` step. This could indicate that some needed items were not created or not given progression classification, or it could indicate that some of the world's manually placemed items (e.g. `location.place_locked_item()`) do not result in a beatable seed with the options that were rolled.

### [item_location_count](https://github.com/Eijebong/Archipelago-fuzzer/blob/main/hooks/item_location_count.py) hook
Fairly simple, this checks that the length of the item pool matches the number of unfilled locations. It's quite rare, but a few Banjo Tooie seeds are putting more items into the item pool than the number of unfilled locations. I can see Tooie logging its own warnings about this in the logs for these hook failures too.

### [indirect_conditions](https://github.com/Eijebong/Archipelago-fuzzer/blob/main/hooks/indirect_conditions.py) hook
This hook basically just runs the `test_explicit_indirect_conditions_spheres` https://github.com/ArchipelagoMW/Archipelago/blob/dc911399fbec508333387439373fdd015663add0/test/general/test_implemented.py#L75-L77 test on fuzzer generations. The failures from this test means that there are missing indirect conditions that affect logic, and could result in nondeterministic generation with a fixed seed.

### [find_missing_indirect_conditions](https://github.com/Mysteryem/Archipelago-fuzzer/blob/mysteryem_hooks/hooks/find_missing_indirect_conditions.py) hook
This is one of my own fuzzer hooks, and attempts to find all cases of `entrance.access_rule()` that check for being able to reach a region without an indirect condition being registered between the entrance and that region. Notably, this hook provides tracebacks to the rules in question, so may be useful to track down the causes of failures from the `indirect_conditions` hook.

Note that this hook can find cases of missing indirect conditions that cannot affect logic, such as if access to Region C is only possible through Region B, and if access to Region B is only possible through Region A (A->B->C), then an Entrance connecting Region A -> Region B that checks for access to Region C, but is missing an indirect condition for Region C can never cause issues because Region B will always end up reachable before Region C becomes reachable.

If a world passes the `indirect_conditions` hook, but fails the `find_missing_indirect_conditions` hook, it could mean that the remaining missing indirect conditions have no effect on logic.

### General FillErrors
These may be caused by previously mentioned issues, so I wouldn't bother looking at them too much, or some may be due to restrictive starts running out of swaps before finding a valid solution. I have included these failures in the zip for completeness. Retrying some of these failed seeds after issues are resolved could result in the seeds succeeding.

g0goTAS commented:
I will look into this, and your PR that improves the performance with the UT glitched logic (we just realised today that it existed).

Before looking at this, I'll point out a couple of issues that we're aware, and that we've decided not to fix, due the amount of work that's needed.

With randomised boss and world loading zones, there's cases where it can lead to unbeatable seeds in very complex and specific situations. With the current way that it's written, we're not aware of an easy to detect those cases, and is something we want to save for a complete rewrite.

With that, I'll take a look at the issues when I have time. I expect to have an update ready by this weekend or the next one.

Mysteryem commented:
> With randomised boss and world loading zones, there's cases where it can lead to unbeatable seeds in very complex and specific situations. With the current way that it's written, we're not aware of an easy to detect those cases, and is something we want to save for a complete rewrite.

You could do similar to what I did when debugging, checking the beatability directly.

Currently, these seeds that are failing, due to an unbeatable seed, can fail generation as late as output, which is not ideal when it could be detected much earlier into a multiworld generation.

I've modified this to cover a single world and assume this check is happening before item plando (checking after item plando would be much more expensive since you then have to sweep the entire multiworld), and added my own commentary on using this function as `##` comments.

```python
    def check_er_beatability(self):
        ## By running this check *before* item plando, it is only necessary to collect the current world's items and
        ## pre-placed items in the current world, which has much better performance scaling in large multiworlds.
        test_state = CollectionState(self.multiworld)
        # Collect our progression items from the item pool.
        for item in self.multiworld.itempool:
            if item.player == self.player and item.advancement:
                test_state.collect(item, True)
        ## This is being run before the `pre_fill` step, so also collect any items that are going to be placed in the
        ## `pre_fill` step.
        # Collect our progression items that are going to be placed in the `pre_fill` step.
        for item in self.get_pre_fill_items():
            if item.advancement:
                test_state.collect(item, True)
        # Collect our reachable pre-placed progression items.
        test_state.sweep_for_advancements(self.get_locations())
        ## If the loading zone randomization is performed after the item pool is created, e.g. in `connect_entrances`,
        ## then accessibility/beatability failures here could be used to perform a limited number of retries at loading
        ## zone randomization, in which case, this function could be changed to return True/False instead of raising
        ## exceptions.
        if not self.multiworld.has_beaten_game(test_state, self.player):
            raise OptionError(f"{self.player_name}: The game is unbeatable with the rolled loading zone randomization.")
        ## Technically, MultiWorld.fulfills_accessibility() only raises an exception with __debug__, so it would be
        ## optional to check accessibility, but note that AP's webhost runs with __debug__=True, so this would fail
        ## there.
        if __debug__:
            ## If you were to opt-in to Items accessibility in the future, that would be a separate check that all your
            ## pre-placed progression items are reachable.
            # All locations must be reachable in Full accessibility.
            if self.options.accessibility == "full":
                for loc in self.get_locations():
                    if not loc.can_reach(test_state):
                        raise OptionError(f"{self.player_name}: The Full Accessibility requirement could not be met"
                                          f" with the rolled loading zone randomization. {loc} could not be reached.")
```

g0goTAS commented:
Sorry if this is taking longer than expected, these bugs are really hard to figure out.

For the location/item mismatches, they all came down to a single cause: the validation for not allowing bass/treble clefs when notes are not randomised was wrong, and all of those yamls should have failed at the yaml validation because of that.

About 10 indirect conditions got added. It's nice that the error logs say which ones need to be added.

For the fill error and accessibility checks. I've been stuck on that, but am making progress. Some of them are inconsistent failures, which I take it as "single-yaml seeds are hard to generate with these options". However, 43 of them are consistent failures, and they all happen with one option. `randomize_worlds: true`, and those are the ones that I'm trying to figure out.

Edit: found the issue. The 43 yamls all have `skip_puzzles: false` and `randomize_worlds: true`, which is a combination of options that should not exist together, and that needs to be added to the validation.

g0goTAS committed:
Fuzzer bugfix

g0goTAS committed:
Merge branch 'main' of https://github.com/jjjj12212/AP-Tooie-PR

g0goTAS committed:
Mumbo logic bugfix

g0goTAS committed:
Faster cryptic hints for large multiworlds

g0goTAS committed:
Merge pull request #3 from Mysteryem/banjo_tooie_faster_non_ut_gen

Improve generation performance by removing UT logic from normal generation

jjjj12212 committed:
Various fixes, optimizations

jjjj12212 committed:
Merge branch 'main' of https://github.com/jjjj12212/AP_Banjo-Tooie

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

jjjj12212 committed:
Merge branch 'ArchipelagoMW:main' into main

ScootyPuffJr1 commented:
We had 31 worlds from this game in the last Big Async and didn't experience any issues whatsoever.
